runtime/lib/backend/openqasm/OpenQasmDevice.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "OpenQasmDevice.hpp"
+
+namespace Catalyst::Runtime::Device {
+
+auto OpenQasmDevice::AllocateQubit() -> QubitIdType
+{
+    RT_FAIL("Unsupported functionality");
+    return 0;
+}
+
+auto OpenQasmDevice::AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType>
+{
+    if (num_qubits == 0U) {
+        return {};
+    }
+
+    const size_t cur_num_qubits = this->device->getNumQubits();
+    const size_t new_num_qubits = cur_num_qubits + num_qubits;
+    if (cur_num_qubits) {
+        this->device = std::make_unique<OpenQasm::OpenQasmBuilder>();
+    }
+
+    this->device->Register(OpenQasm::RegisterType::Qubit, "qubits", new_num_qubits);
+
+    return this->qubit_manager.AllocateRange(cur_num_qubits, new_num_qubits);
+}
+
+void OpenQasmDevice::ReleaseAllQubits()
+{
+    // do nothing
```

Comment:
When does this function get called from other devices? It looks like the decomposition is the following:

1. JAXPR qdealloc
2. quantum.dealloc
3. @__quantum__rt__qubit_release_array
4. ReleaseAllQubits

It looks like we do allocate qubits just above, but we don't deallocate them like we do for other devices. Why is this?

Reply:
This is because of the fact that the deallocations of qubits aren't supported for OpenQasm programs passing through Amazon Braket SDK. The deallocation support is implemented in the `OpenQasmBuilder` class though but not used in this method.